tests/Integration/ValidationCachingTest.php

```diff
@@ -0,0 +1,250 @@
+            ],
+        ]);
+
+        Event::assertDispatchedTimes(CacheMissed::class, 0);
+        Event::assertDispatchedTimes(CacheHit::class, 0);
+        Event::assertDispatchedTimes(KeyWritten::class, 0);
+    }
+
+    public function testErrorsAreNotCached(): void
+    {
+        $config = $this->app->make(ConfigRepository::class);
+        $config->set('lighthouse.query_cache.enable', false);
+        $config->set('lighthouse.validation_cache.enable', true);
+
+        Event::fake();
+
+        $this->graphQL(/** @lang GraphQL */ '
+        {
+            bar
+        }
+        ')->assertGraphQLErrorMessage('Cannot query field "bar" on type "Query".');
+
+        Event::assertDispatchedTimes(CacheMissed::class, 1);
+        Event::assertDispatchedTimes(CacheHit::class, 0);
+        Event::assertDispatchedTimes(KeyWritten::class, 0);
+    }
+
+    public function testDifferentQueriesHasDifferentKeys(): void
+    {
+        $config = $this->app->make(ConfigRepository::class);
+        $config->set('lighthouse.query_cache.enable', false);
+        $config->set('lighthouse.validation_cache.enable', true);
+
+        Event::fake();
+
+        $this->graphQL(/** @lang GraphQL */ '
+        {
+            foo
+        }
+        ')->assertExactJson([
+            'data' => [
+                'foo' => Foo::THE_ANSWER,
+            ],
+        ]);
+
+        $this->graphQL(/** @lang GraphQL */ '
+        {
+            foo
+            foo
+        }
+        ')->assertExactJson([
+            'data' => [
+                'foo' => Foo::THE_ANSWER,
+            ],
+        ]);
+
+        Event::assertDispatchedTimes(CacheMissed::class, 2);
+        Event::assertDispatchedTimes(CacheHit::class, 0);
+        Event::assertDispatchedTimes(KeyWritten::class, 2);
+    }
+
+    public function testSameSchemaAndSameQueryHaveSameKeys(): void
+    {
+        $config = $this->app->make(ConfigRepository::class);
+        $config->set('lighthouse.query_cache.enable', false);
+        $config->set('lighthouse.validation_cache.enable', true);
+
+        $event = Event::fake();
+
+        $this->graphQL(/** @lang GraphQL */ '
+        {
+            foo
+        }
+        ')->assertExactJson([
+            'data' => [
+                'foo' => Foo::THE_ANSWER,
+            ],
+        ]);
+
+        $event->assertDispatchedTimes(CacheMissed::class, 1);
+        $event->assertDispatchedTimes(CacheHit::class, 0);
+        $event->assertDispatchedTimes(KeyWritten::class, 1);
+
+        // refresh container, but keep the same cache
+        $cacheFactory = $this->app->make(CacheFactory::class);
+        $this->refreshApplication();
+        $this->app->instance(CacheFactory::class, $cacheFactory);
+
+        $this->setUpTestSchema();
+
+        $config = $this->app->make(ConfigRepository::class);
+        $config->set('lighthouse.query_cache.enable', false);
+        $config->set('lighthouse.validation_cache.enable', true);
+
+        Event::fake();
```

Comment:
why fake twice?

Reply:
To override the default event handler after application refresh. Changed it to `$this->app->instance(EventsDispatcher::class, $event);` so the new application would get the same event dispatcher.